sodetlib/operations/bias_wave.py

```diff
@@ -0,0 +1,565 @@
+                each bias group
+            dVbias:
+                Array of shape (nbgs) containing the wave voltage amplitude for
+                each bias group
+            dItes:
+                Array of shape (nchans) containing the wave current amplitude for
+                each detector.
+        """
+        nbgs = 12
+        nchans = len(self.am.signal)
+        npts = np.nanmin(self.stop_idxs-self.start_idxs)
+
+        Ibias = np.full(nbgs, np.nan)
+        dIbias = np.full(nbgs, 0.0, dtype=float)
+        dItes = np.full(nchans, np.nan)
+
+        # Compute Ibias and dIbias
+        bias_line_resistance = self.meta['bias_line_resistance']
+        high_low_current_ratio = self.meta['high_low_current_ratio']
+        rtm_bit_to_volt = self.meta['rtm_bit_to_volt']
+        amp_per_bit = 2 * rtm_bit_to_volt / bias_line_resistance
+        if self.high_current_mode:
+            amp_per_bit *= high_low_current_ratio
+        
+        for bg in range(nbgs):
+            if len(self.start_idxs[bg]) == 0:
+                continue
+            rcs = np.where(self.bgmap == bg)[0]
+            s = slice(int(self.start_idxs[bg][0]),
+                      int(self.start_idxs[bg][0] + npts))
+            Ibias[bg] = np.nanmean(self.am.biases[bg, s]) * amp_per_bit
+            dIbias[bg] = get_amplitudes(self.run_kwargs['freqs_wave'][0],
+                                        self.am.biases[bg, s])[0] * amp_per_bit
+            dItes[rcs] = get_amplitudes(self.run_kwargs['freqs_wave'][0],
+                                        self.wave_resp[rcs,0,:])
+
+        self.Ibias = Ibias
+        self.Vbias = Ibias * bias_line_resistance
+        self.dIbias = dIbias
+        self.dVbias = dIbias * bias_line_resistance
+        self.dItes = dItes
+
+        R0, I0, Pj = self._compute_R0_I0_Pj()
+
+        Si = -1./(I0 * (R0 - self.meta['R_sh']))
+
+        # If resistance is too high, most likely crosstalk so just reset
+        # bg mapping and det params
+        if R0_thresh is not None:
+            m = np.abs(R0) > R0_thresh
+            self.bgmap[m] = -1
+            for arr in [R0, I0, Pj, Si]:
+                arr[m] = np.nan
+
+        self.R0 = R0
+        self.I0 = I0
+        self.Pj = Pj
+        self.Si = Si
+
+        return R0, I0, Pj, Si
+
+    def _compute_R0_I0_Pj(self):
+        """
+        Computes the DC params R0 I0 and Pj
+
+        Carbon copy from BiasStepAnalysis
+        """
+        Ib = self.Ibias[self.bgmap]
+        dIb = self.dIbias[self.bgmap]
+        dItes = self.dItes
+        Ib[self.bgmap == -1] = np.nan
+        dIb[self.bgmap == -1] = np.nan
+        dIrat = dItes / dIb
+    
+        R_sh = self.meta["R_sh"]
+
+        I0 = np.zeros_like(dIrat)
+        I0_nontransition = Ib * dIrat
+        I0_transition = Ib * dIrat / (2 * dIrat - 1)
+        I0[dIrat>0] = I0_nontransition[dIrat>0]
+        I0[dIrat<0] = I0_transition[dIrat<0]
+
```

Comment:
That's true...we need a way to check the phase of dItes relative to dIbias (would be easier with directly projecting out the frequency instead of taking from the periodogram)

Reply:
I vaguely recall Yuhan and I thinking about this a while back... I'll talk with him once he has a bit of time and I can refresh on what we discussed here/potentially change this

Reply:
I just committed some changes meant to fix this with the help of Max. Let me know what y'all think.

Reply:
Did you test it and see that you got out negative values when you expected?

Reply:
Maybe I'm misunderstanding... but I don't think the changes would work. Phase info isn't included in dItes or dIbias (since they're computed via sqrt(Pxx)), so I don't think its possible to extract the phase like this.

Reply:
Have corrected calculation of amplitudes to allow for distinction between in- and out- of transition in recent commits.